frontends/concrete-python/examples/sha1/sha1.md

```diff
@@ -0,0 +1,1726 @@
+    def rotate5(x):
+        ans = fhe.zeros((32,))
+        ans[5:32] = x[0:27]
+        ans[0:5] = x[27:32]
+        return ans
+
+    @fhe.function({"x": "encrypted", "y": "encrypted"})
+    def add2(x, y):
+        ans = fhe.zeros((32,))
+        cy = 0
+
+        for i in range(32):
+            t = x[i] + y[i] + cy
+            cy, tr = t >= 2, t % 2
+            ans[i] = tr
+
+        return ans
+
+    @fhe.function(
+        {"x": "encrypted", "y": "encrypted", "u": "encrypted", "v": "encrypted", "w": "encrypted"}
+    )
+    def add5(x, y, u, v, w):
+        ans = fhe.zeros((32,))
+        cy = 0
+
+        for i in range(32):
+            t = x[i] + y[i] + cy
+            cy, tr = t // 2, t % 2
+            ans[i] = tr
+
+        cy = 0
+
+        for i in range(32):
+            t = ans[i] + u[i] + cy
+            cy, tr = t // 2, t % 2
+            ans[i] = tr
+
+        cy = 0
+
+        for i in range(32):
+            t = ans[i] + v[i] + cy
+            cy, tr = t // 2, t % 2
+            ans[i] = tr
+
+        cy = 0
+
+        for i in range(32):
+            t = ans[i] + w[i] + cy
+            cy, tr = t // 2, t % 2
+            ans[i] = tr
+
+        return ans
+```
+
+It contains 7 functions which can be combined together:
+- `xor3` XORs three values together
+- `iftern` computes the ternary IF, i.e., c ? t : f
+- `maj` computes the majority function
+- `rotate30` rotates a 32-bit word by 30 positions
+- `rotate5` rotates a 32-bit word by 5 positions
+- `add2` adds two values together
+- `add5` adds five values together
+
+This module is then compiled. Remark we set `p_error=10**-8` to a very small value, to avoid
+errors in computations. Thanks to the Module computation, all these functions can be combined
+together, i.e., outputs of some of them can be reused as inputs of some other ones. It is very
+convenient to have a larger function, with some control-flow (conditions, branches, loops) done in
+the clear and operating with these small functions. In our case, it's done in
+`_process_encrypted_chunk_server_side` function.
+
+## Details of `_process_encrypted_chunk_server_side`
+
+`_process_encrypted_chunk_server_side` uses inputs which are encrypted, and returns encrypted
+values. Everything is a 32-bit word when done in the clear. Here, words will be represented as 32
+1-bit encrypted values, to simplify (and speed) the non-linear operations in SHA1.
+
+Then, we have the main loop of the compression function which is done in the clear with
+`for i in range(80):`. There, we also have the different conditions (`if 0 <= i <= 19:`,
+`elif 20 <= i <= 39` etc) which is done in the clear. Obviously, there is no security issue here,
+since the `i` index is not secret. Depending on `i`, we then select the right `fsplit_enc` and
+`ksplit`, which correspond to `f` and `k` in the clear implementation of the compression function.
+Here, we can see a first use of the different functions in the Module.
```

Comment:
Isn't it possible to embed the code in the md file directly ? Isn't it asking too much of users to kind of switch from the markdown to the file ? The fact that you refer to an external file makes it pretty different from the tutorials I am used to, where the authors generally build the code as they explain things.

Reply:
sure we can but I guess, I wanted to explain, instead of just showing
but at the same time, what I do is paraphrasing some code which is pretty trivial to read, so I will do as you say

Reply:
Isn't it possible to integrate jupyter notebooks in our doc ? Some of the best tutorials I read use them ...

Reply:
it is, but I don't personally like the fact that in Jupyter, you can run cell A, then cell B, then cell A again etc; and depending on the order of your execution, results may vary. To me, there is nothing better than a .py to launch.

Reply:
If the team votes for a Jupyter, I'll do it. 

Even measuring perfs on Jupyter: I'm not sure execution times are as fast as in pure python.

Reply:
for this part, I wait for other inputs, eg @BourgerieQuentin

Reply:
I ok for a raw python program, but I'm ok with @aPere3 it's boring to switch. I prefer to have code inlined here.

Reply:
Sure, adding inlined code, you'll tell me if it's sufficient